extensions/aws/processors/PutS3Object.cpp

```diff
@@ -0,0 +1,487 @@
+    logger_->log_error("Canned ACL is invalid!");
+    return false;
+  }
+  logger_->log_debug("PutS3Object: Canned ACL [%s]", put_s3_request_params_.canned_acl);
+  return true;
+}
+
+bool PutS3Object::setAccessControl(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::FlowFile> &flow_file) {
+  std::string value;
+  if (context->getProperty(FullControlUserList, value, flow_file) && !value.empty()) {
+    put_s3_request_params_.fullcontrol_user_list = parseAccessControlList(value);
+    logger_->log_debug("PutS3Object: Full Control User List [%s]", value);
+  }
+  if (context->getProperty(ReadPermissionUserList, value, flow_file) && !value.empty()) {
+    put_s3_request_params_.read_permission_user_list = parseAccessControlList(value);
+    logger_->log_debug("PutS3Object: Read Permission User List [%s]", value);
+  }
+  if (context->getProperty(ReadACLUserList, value, flow_file) && !value.empty()) {
+    put_s3_request_params_.read_acl_user_list = parseAccessControlList(value);
+    logger_->log_debug("PutS3Object: Read ACL User List [%s]", value);
+  }
+  if (context->getProperty(WriteACLUserList, value, flow_file) && !value.empty()) {
+    put_s3_request_params_.write_acl_user_list = parseAccessControlList(value);
+    logger_->log_debug("PutS3Object: Write ACL User List [%s]", value);
+  }
+
+  return setCannedAcl(context, flow_file);
+}
+
+bool PutS3Object::getExpressionLanguageSupportedProperties(
+    const std::shared_ptr<core::ProcessContext> &context,
+    const std::shared_ptr<core::FlowFile> &flow_file) {
+  context->getProperty(ObjectKey, put_s3_request_params_.object_key, flow_file);
+  logger_->log_debug("PutS3Object: Object Key [%s]", put_s3_request_params_.object_key);
+  if (!context->getProperty(Bucket, put_s3_request_params_.bucket, flow_file) || put_s3_request_params_.bucket.empty()) {
+    logger_->log_error("Object Key is invalid or empty!", put_s3_request_params_.bucket);
+    return false;
+  }
+  logger_->log_debug("PutS3Object: Bucket [%s]", put_s3_request_params_.bucket);
+
+  context->getProperty(ContentType, put_s3_request_params_.content_type, flow_file);
+  logger_->log_debug("PutS3Object: Content Type [%s]", put_s3_request_params_.content_type);
+
+  auto credentials = getAWSCredentials(context, flow_file);
+  if (!credentials) {
+    logger_->log_error("AWS Credentials have not been set!");
+    return false;
+  }
+  s3_wrapper_->setCredentials(credentials.value());
+
+  if (!setProxy(context, flow_file)) {
+    context->yield();
+    return false;
+  }
+
+  std::string value;
+  if (context->getProperty(EndpointOverrideURL, value, flow_file) && !value.empty()) {
+    s3_wrapper_->setEndpointOverrideUrl(value);
+    logger_->log_debug("PutS3Object: Endpoint Override URL [%d]", value);
+  }
+
+  return setAccessControl(context, flow_file);
+}
+
+void PutS3Object::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
```

Comment:
some processors are now receiving a `Batch Size` property to consume more flowFiles (but at most `Batch Size`) in a single trigger, this processor looks like a good candidate for such a behavior, it could also be implemented later in a follow-up PR, what do you think?

Reply:
I think it's a good idea as an upload could take a some time depending on the latency/region/object size so the scheduling and load balancing could be optimized with this. I'm still discussing some requirements with @arpadboda I'll check with him about this as well.

Reply:
I discussed with @arpadboda and we do not really have any benefit of the batch processing in case of an `EVENT_DRIVEN` processor (which should be a main use case) as we still have to do every processing step on a single flow file basis. Unfortunately [S3 does not support batch upload of objects](https://stackoverflow.com/questions/15050146/is-it-possible-to-perform-a-batch-upload-to-amazon-s3) so cannot benefit from a single request batch upload either.